### PR TITLE
Mechanism to check rosdep rules against package repositories

### PR DESCRIPTION
**Pull request by cottsay**
This change adds a small system for checking rosdep rules against the current package repositories for supported platforms.

It currently supports three platforms:
- Ubuntu
- Fedora
- RHEL

During a typical rosdep test invocation, it uses a similar mechanism as the one which verifies URLs to detect which lines were added or modified by the latest change, and reviews only those rules.

The algorithm is optimized such that the sources are listed in the order they are checked for the package, and parsing of the repository metadata begins before the download is even completed. This means that presence in the platform repositories can be determined very quickly.

If multiple package names were added or modified, the algorithm first checks the data that has already been parsed and then continues parsing metadata until the package is found or the repository list is exhausted.

To review the entire rosdep database and not only the rules that were changed in the most recent commit, it is also possible to invoke the test as a python module. For example:

`PYTHONPATH=test python3 -m rosdep_repo_check`

This system is currently identifying many missing packages throughout rosdep. We should review these issues at some point, but this system should help to prevent more problematic rules from being added.

### CODE REVIEW

**Comment by cottsay** [comment]
Some ideas to improve upon what's presented here:
* Support more platforms. It should be pretty easy to add more.
* More robust logic for retrying repository metadata downloads. The existing logic doesn't work very well with the progressive downloading strategy because HTTP errors won't manifest until data is received, which is after the handle has been returned from the URL opening function. Maybe wait for the headers somehow?
* Support for RPM mirrorlists and/or metalinks.
* Express suggestions by creating PR review and actually suggestion the change in the code, rather than throwing a build warning that's hard to discover.
* Teach suggestion code how to check more OS versions than just the newest. Right now, it could make a suggestion that is only valid for the latest OS version and, if committed, would fail the repository check.
* Better rosdep parsing logic to handle rules which specify the package manager. Right now, it just assumes everything that's not `pip` is a system package that should be checked against the configured repository. This is a largely safe assumption, but still isn't correct.

**Review by hidmic — COMMENTED** [review]
I made a pass. Code looks great, I don't have much to add. 

I do see what seems like heavy heuristics to me (or a very diverse set of conventions). How stable do you think we can expect this code to be?

Also, some documentation for functions and classes would be nice. A few comments here and there explaining the nuisances (like e.g. how the annotated YAML loader causes the OS version to end up being equal to `__line__`) would help readability a lot (for me at least).

**Review by j-rivero — APPROVED** [review]
Far from being an expert in rosdep but the code and the test looks good to me.

File: test/rosdep_repo_check/test_rosdep_repo_check.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+def detect_lines(diffstr):
+    """Take a diff string and return a dict of
+    files with line numbers changed"""
+    resultant_lines = {}
+    # diffstr is already decoded
+    io = StringIO(diffstr)
+    udiff = unidiff.PatchSet(io)
+    for file in udiff:
+        target_lines = []
+        # if file.path in TARGET_FILES:
```
[the remark this answers is not in the document]
Comment (cottsay) [reply test/rosdep_repo_check/test_rosdep_repo_check.py]: 3b86bbd

File: test/rosdep_repo_check/__init__.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            sources = os_sources.get(os_code_name, [])
+        else:
+            sources = [os_sources]
+        if not sources:
+            print('WARNING: No sources for %s %s' % (os_name, os_code_name), file=sys.stderr)
+        for source in sources:
+            for p in [
+                    p for p in source.enumerate_packages(os_name, os_code_name, os_arch)
+                    if p == pkg_name]:
+                return p
```
[the remark this answers is not in the document]
Comment (cottsay) [reply test/rosdep_repo_check/__init__.py]: 3b86bbd

**Review by nuclearsandwich — APPROVED** [review]
This is very awesome!

~~I added a basic README for this module~~ (didn't realize this PR is hosted on a fork), mostly adapting information from the pull request. I have also started fooling around adding opensuse and debian checks as well as README info about how to update or add new platforms but they're not ready yet and I don't want to block this.

---

It's a bit unfortunate that we have to create an additional list of supported platforms for just this script and I started down the path of trying to read platforms from currently active rosdistros but nothing is architecture specific without reading from ros_buildfarm_config as well. On that note are we concerned that we're only checking amd64/x86_64 ports and not other architectures?
On top of all that limiting these checks to just released platforms prevents us from checking community supported platforms like opensuse.

**Comment by cottsay** [comment]
> I started down the path of trying to read platforms from currently active rosdistros

This is a great idea, and I hadn't thought about doing that.

> I added a basic README for this module, mostly adapting information from the pull request. I have also started fooling around adding opensuse and debian checks as well as README info about how to update or add new platforms but they're not ready yet and I don't want to block this.

Superb.

> are we concerned that we're only checking amd64/x86_64 ports and not other architectures?

I'm not excluding architectures on purpose. Feel free to try enabling more. It's pretty much a direct trade-off with runtime, because it needs to pull more repository metadata. If it starts to get unreasonable, we should consider threading the whole works.

> On top of all that limiting these checks to just released platforms prevents us from checking community supported platforms like opensuse.

Fair point. Maybe making a hybrid definition format is appropriate - grab the supported stuff automatically from the rosdistro index, and supplement that with a curated list of community platforms.

**Comment by dedi96irawan** [comment]
https://youtube.com/channel/UCihdVEwTrggEo4Ho_Iqs_PQ

⁣Get TypeApp for Android ​

**Pull request merged**